pep-0xxx.rst

```diff
@@ -0,0 +1,1250 @@
+            from ``incoming`` and decrypt it, and write encrypted data to
+            ``outgoing``.
+
+            The buffer objects must be either file objects or objects that
+            implement the buffer protocol.
+            """
+
+
+Socket
+~~~~~~
+
+The socket-wrapper ABC will be defined by the ``TLSWrappedSocket`` ABC, which
+has the following definition::
+
+    class TLSWrappedSocket(metaclass=ABCMeta):
+        # The various socket methods all must be implemented. Their definitions
+        # have been elided from this class defintion in the PEP because they
+        # aren't instructive.
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[CipherSuite]:
+            """
+            Returns the CipherSuite entry for the cipher that has been
+            negotiated on the connection. If no connection has been negotiated,
+            returns ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[Union[NextProtocol, bytes]]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If the negotiated protocol is one of the protocols defined in the
+            ``NextProtocol`` enum, the value from that enum will be returned.
+            Otherwise, the raw bytestring of the negotiated protocol will be
+            returned.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def unwrap(self) -> socket.socket:
+            """
+            Cleanly terminate the TLS connection on this wrapped socket. Once
+            called, this ``TLSWrappedSocket`` can no longer be used to transmit
+            data. Returns the socket that was wrapped with TLS.
+            """
+
+
+Buffer
+~~~~~~
+
+The buffer-wrapper ABC will be defined by the ``TLSWrappedBuffer`` ABC, which
+has the following definition::
+
+    class TLSWrappedBuffer(metaclass=ABCMeta):
```

Comment:
Yeah, that could work. I'm inclined to want to tag in @glyph here, as he's had some experience with OpenSSL's BIO objects, and so probably has a good idea of what kind of interface would be more useful.

@glyph, rather than reading the whole discussion, I'll summarise here. Which of the following forms would be more useful for a non-blocking TLS API?

1. Have an object with a `read()` method that takes a collection of bytes and decodes them as it is able to, returning the decoded bytes and buffering the rest; and a `write()` method that takes cleartext bytes and encodes as many as possible, buffering the rest, and returning the ciphertext bytes?
2. Having a `read()` method that reads from an internal buffer of ciphertext and returning what it can, a `write()` method that writes to an internal buffer of ciphertext, and then a method that allows writing bytes from the network into the ciphertext buffer and one that allows reading bytes from the ciphertext buffer out to the network?